xapian-core/include/xapian/intrusive_ptr.h

```diff
@@ -236,7 +236,33 @@ template<class T> class intrusive_ptr_nonnull
 
     ~intrusive_ptr_nonnull()
     {
-        if(--px->_refs == 0 ) delete px;
+        if(px != 0 && --px->_refs == 0 ) delete px;
+    }
+
+    intrusive_ptr_nonnull(intrusive_ptr_nonnull && rhs) : px( rhs.px )
+    {
+        rhs.px = 0;
```

Comment:
@ojwb Here for source object invariant is lost, but the object is going to be destructed. That's ok right?

Reply:
That's an interesting question.

In typical use the object will soon be destroyed, but that's not a requirement when using `std::move()`.

http://en.cppreference.com/w/cpp/language/move_constructor says:

> Move constructors typically "steal" the resources held by the argument (e.g. pointers to dynamically-allocated objects, file descriptors, TCP sockets, I/O streams, running threads, etc.) rather than make copies of them, and leave the argument in some valid but otherwise indeterminate state.

So that suggests it's not conventional to leave the object in an invalid state like this.

I think it's probably safer to just increment the reference count here (i.e. make it do the same as the copy constructor).  That's not a big overhead, and the compiler might even be able to optimise away the increment and decrement if the object is immediately destroyed.

Your move assignment (which essentially does `swap()`) is fine, as that preserves the invariant.

Reply:
You are right, thanks! code is updated.